Terminal.Gui/View/ViewScrollBar.cs

```diff
@@ -0,0 +1,635 @@
+﻿namespace Terminal.Gui;
+
+/// <summary>
+///     The scroll bar types used by this <see cref="ScrollBarView"/>.
+/// </summary>
+public enum ScrollBarType
+{
+    /// <summary>
+    ///     None scroll bar will be used and to avoid throwing an exception never use it in a constructor.
+    /// </summary>
+    None,
+
+    /// <summary>
+    ///     Only the vertical scroll bar will be shown.
+    /// </summary>
+    Vertical,
+
+    /// <summary>
+    ///     Only the horizontal scroll bar will be shown.
+    /// </summary>
+    Horizontal,
+
+    /// <summary>
+    ///     Both vertical and horizontal scroll bars will be shown.
+    /// </summary>
+    Both
+}
+
+public partial class View
+{
+    private ScrollBarView _scrollBar;
+    private ScrollBarType _scrollBarType;
+    private int _scrollColsSize;
+    private int _scrollLeftOffset;
+    private int _scrollRowsSize;
+    private int _scrollTopOffset;
+
+    /// <summary>If true the vertical/horizontal scroll bars won't be showed if it's not needed.</summary>
+    public bool ScrollAutoHideScrollBars
+    {
+        get => _scrollBar.AutoHideScrollBars;
+        set => _scrollBar.AutoHideScrollBars = value;
+    }
+
+    /// <summary>
+    ///     Gets or sets the <see cref="ScrollBarType"/> used by this view.
+    /// </summary>
+    /// <exception cref="ArgumentOutOfRangeException">The value is out of range.</exception>
+    public virtual ScrollBarType ScrollBarType
+    {
+        get => _scrollBarType;
+        set
+        {
+            if (_scrollBar is { } && _scrollBarType == value)
+            {
+                return;
+            }
+
+            _scrollBarType = value;
+            DisposeScrollBar ();
+
+            switch (_scrollBarType)
+            {
+                case ScrollBarType.Vertical:
+                    _scrollBar = new ScrollBarView { IsVertical = true };
+
+                    break;
+                case ScrollBarType.Horizontal:
+                    _scrollBar = new ScrollBarView { IsVertical = false };
+
+                    break;
+                case ScrollBarType.Both:
+                    _scrollBar = new ScrollBarView { IsVertical = true };
+                    _scrollBar.OtherScrollBarView = new ScrollBarView { IsVertical = false, OtherScrollBarView = _scrollBar };
+
+                    break;
+                case ScrollBarType.None:
+                    return;
+                default:
+                    throw new ArgumentOutOfRangeException ();
+            }
+
+            Padding.Add (_scrollBar);
+            AddHandlers ();
+            AddScrollKeyBindings (_scrollBar);
+
+            if (_scrollBar.OtherScrollBarView != null)
```

Comment:
I don't recommend disconnecting them because one is always dependent on the other for adjustments so that they are always synchronized. The only care to be taken is not to add additional views. But another method of achieving the same objective may also be debatable.

Reply:
Please inventory all of the places/scenarios where you think they are "always dependent on the other". I believe that list will be very short, and easily addressed without such a tight coupling.

I believe (I could be wrong, so please educate me if so) that the coupling that exists today is there because that's how it was originally authored. I believe this is overly complex and can be simplified rather easily.

Reply:
I am referring to the case in which `KeepContentOnViewPort` is true and a scroll bar is invisible and in this case there is another vacant unit to accommodate this space. If there is no interconnection mechanism, the other scroll bar will not adjust to accommodate this available space. Iterating between them is much simpler if they can communicate with each other.